modules/core/osgi-service-tracker-collections/src/testIntegration/java/com/liferay/osgi/service/tracker/collections/map/test/ListServiceTrackerMapTest.java

```diff
@@ -300,6 +300,81 @@ public void testGetServiceWithSimpleRegistrationTwice() {
 		serviceTrackerMap.close();
 	}
 
+	@Test
+	public void testMissingVsSpecifiedServiceRanking() {
+		try (ServiceTrackerMap<String, List<TrackedOne>> serviceTrackerMap =
+				createServiceTrackerMap(_bundleContext)) {
+
+			Dictionary<String, Object> properties1 = new Hashtable<>();
+			Dictionary<String, Object> properties2 = new Hashtable<>();
+
+			properties1.put("target", "aTarget");
+
+			properties2.put("target", "aTarget");
+			properties2.put("service.ranking", Integer.MAX_VALUE);
+
+			TrackedOne service1 = new TrackedOne();
+			TrackedOne service2 = new TrackedOne();
+
+			ServiceRegistration<TrackedOne> serviceRegistration1 =
+				_bundleContext.registerService(
+					TrackedOne.class, service1, properties1);
+
+			ServiceRegistration<TrackedOne> serviceRegistration2 =
+				_bundleContext.registerService(
+					TrackedOne.class, service2, properties2);
+
+			long serviceId1 =
+				(long)serviceRegistration1.getReference().getProperty(
+					"service.id");
+
+			long serviceId2 =
+				(long)serviceRegistration2.getReference().getProperty(
+					"service.id");
+
+			Assert.assertTrue(serviceId1 < serviceId2);
+
+			List<TrackedOne> services = serviceTrackerMap.getService("aTarget");
+
+			Assert.assertEquals(service2, services.get(0));
+			Assert.assertEquals(service1, services.get(1));
+
+			serviceRegistration2.unregister();
+
+			services = serviceTrackerMap.getService("aTarget");
+
+			Assert.assertEquals(service1, services.get(0));
+
+			serviceRegistration1.unregister();
+
+			// Ensure that ordering is not being done by ascending service.id
```

Comment:
Have you entertained the possibility of modifying the properties of the ServiceRegistration and change the `service.ranking` from MAX_VALUE to MIN_VALUE and recheck the assertion? This way you only have to register once.

Reply:
Interesting idea. I didn't know/think that the ServiceReference object is actually the service reference used by the OSGi runtime and updating it will have an effect on the runtime. Learn something every day 👍 

I'll explore this approach.

Reply:
you have to use `ServiceRegistration.setProperties`